### PR TITLE
Refactor threading in arrow-effects for simplicity

### PR DESCRIPTION
**Pull request by pakoito**
4th attempt

### CODE REVIEW

File: modules/core/arrow-annotations-processor/src/main/java/arrow/tc/TypeclassFileGenerator.kt [hunk 1]
```diff
@@ -116,6 +121,17 @@ data class SyntaxFunctionSignature(
                     isAbstract = isAbstract
             )
         }
+
+        // At this point the suspended parameter has already been rewritten and needs to be reverted
+        private fun extractSuspendedArgType(it: ProtoBuf.ValueParameter, typeClass: Typeclass): String {
```
Comment (raulraja) [inline modules/core/arrow-annotations-processor/src/main/java/arrow/tc/TypeclassFileGenerator.kt]: :O

File: modules/core/arrow-mtl/src/main/kotlin/arrow/mtl/MonadFilter.kt [hunk 2]
```diff
@@ -14,16 +20,17 @@ interface MonadFilter<F> : Monad<F>, FunctorFilter<F>, TC {
...
-    val continuation = MonadFilterContinuation<F, B>(this)
-    val wrapReturn: suspend MonadFilterContinuation<F, *>.() -> Kind<F, B> = { pure(c()) }
-    wrapReturn.startCoroutine(continuation, continuation)
-    return continuation.returnedMonad()
+    /**
+     * Entry point for monad bindings which enables for comprehension. The underlying impl is based on coroutines.
+     * A coroutine is initiated and inside [MonadContinuation] suspended yielding to [flatMap]. Once all the flatMap binds are completed
+     * the underlying monad is returned from the act of executing the coroutine
+     */
+    fun <B> bindingFilter(cc: CoroutineContext = EmptyCoroutineContext, c: suspend BindingFilterContinuation<F, *>.() -> B): Kind<F, B> {
```
Comment (raulraja) [inline modules/core/arrow-mtl/src/main/kotlin/arrow/mtl/MonadFilter.kt]: :+1:

File: modules/core/arrow-test/src/main/kotlin/arrow/test/laws/MonadLaws.kt [hunk 3]
```diff
@@ -102,8 +101,8 @@ object MonadLaws {
     inline fun <reified F> monadComprehensionsBindInContext(M: Monad<F> = monad<F>(), EQ: Eq<Kind<F, Int>>): Unit =
             forFew(5, genIntSmall(), { num: Int ->
                 M.binding {
-                    val a = bindIn(newSingleThreadContext("$num")) { num + 1 }
-                    val b = bindIn(newSingleThreadContext("$a")) { a + 1 }
+                    val a = binding(newSingleThreadContext("$num")) { num + 1 }.bind()
```
Comment (raulraja) [inline modules/core/arrow-test/src/main/kotlin/arrow/test/laws/MonadLaws.kt]: :+1:

File: modules/effects/arrow-effects/src/main/kotlin/arrow/effects/continuations/IOContinuations.kt [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+import arrow.core.Either
+import arrow.effects.ForIO
+import arrow.effects.IOMonadErrorInstance
+import arrow.effects.IOOf
+import arrow.effects.fix
+import arrow.typeclasses.continuations.BindingCatchContinuation
+import kotlin.coroutines.experimental.CoroutineContext
+import kotlin.coroutines.experimental.suspendCoroutine
+
+class IOContinuation<A>(private val cb: (Either<Throwable, A>) -> Unit, override val context: CoroutineContext) :
```
Comment (raulraja) [inline modules/effects/arrow-effects/src/main/kotlin/arrow/effects/continuations/IOContinuations.kt]: :+1:

File: modules/core/arrow-mtl/src/main/kotlin/arrow/mtl/continuations/MonadFilterContinuation.kt [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+    override fun resumeWithException(exception: Throwable) {
+        when (exception) {
+            PredicateInterrupted -> returnedMonad = empty()
+            else -> super.resumeWithException(exception)
+        }
+    }
+
+    override fun continueIf(predicate: Boolean) {
+        if (!predicate) throw MonadFilterContinuation.PredicateInterrupted
```
Comment (nomisRev) [inline modules/core/arrow-mtl/src/main/kotlin/arrow/mtl/continuations/MonadFilterContinuation.kt]: ELI5 😁 Can you not immediately do the assignment here? `returnedMonad = empty()`

File: modules/core/arrow-typeclasses/src/main/kotlin/arrow/typeclasses/Monad.kt [hunk 6]
```diff
@@ -27,16 +32,17 @@ interface Monad<F> : Applicative<F>, TC {
...
- * the underlying monad is returned from the act of executing the coroutine
- */
-fun <F, B> Monad<F>.binding(c: suspend MonadContinuation<F, *>.() -> B): Kind<F, B> {
-    val continuation = MonadContinuation<F, B>(this)
-    val wrapReturn: suspend MonadContinuation<F, *>.() -> Kind<F, B> = { pure(c()) }
-    wrapReturn.startCoroutine(continuation, continuation)
-    return continuation.returnedMonad()
+    /**
+     * Entry point for monad bindings which enables for comprehension. The underlying implementation is based on coroutines.
+     * A coroutine is initiated and suspended inside [MonadErrorContinuation] yielding to [Monad.flatMap]. Once all the flatMap binds are completed
```
Comment (nomisRev) [inline modules/core/arrow-typeclasses/src/main/kotlin/arrow/typeclasses/Monad.kt]: `[BindingContinuation]`?

Comment (pakoito) [reply modules/core/arrow-typeclasses/src/main/kotlin/arrow/typeclasses/Monad.kt]: All the docs are outdated. I have to clean them up before merging this. WIP!

File: modules/core/arrow-mtl/src/main/kotlin/arrow/mtl/continuations/MonadFilterContinuation.kt (hunk 5 above)
Comment (pakoito) [reply modules/core/arrow-mtl/src/main/kotlin/arrow/mtl/continuations/MonadFilterContinuation.kt]: I didn't write this, so I'll defer to the original creator xD

**Comment by raulraja** [comment]
@pakoito is this still relevant?

**Comment by pakoito** [comment]
@raulraja some parts of it still are, like replacing MonadContinuations with an interface. I kept it around for reference.

The parts around threading are unnecessary after @JorgeCastilloPrz and @nomisRev 's changes.

**Comment by pakoito** [comment]
`Async<E, A>` is still worth considering, specially after we get a BIO.

**Comment by nomisRev** [comment]
@pakoito can we close this?
